Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,335 @@
+
+    if hash(address) % shards_per_zone == assignment_idx {
+        do_scrape()
+    }
+}
+```
+
+By using modulo to calculate the `prom_zone_idx`, instances will be distributed
+to zones in the sense of `A,B,C,A,B,C` and so on. This allows a modification of
+the `num_shards` value without redistribution of shards or data.
+This was preferred over allowing the number of `zones` to change, as this is
+less likely to happen.
+
+### Edge cases
+
+We have introduced asserts in the above section to warn about edge cases that
+might lead to duplicate data or data loss.
+
+By the above algorithm. prometheus instances will be distributed in an
+alternating fashion by using the already existing shard index.
+This leads to the following edge cases:
+
+When `num_shards` is 10 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 3. This yields the following distribution
+
+```text
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```text
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
+
+This second case - a zone not being scraped - should lead to an error in the
+operator, causing the change to not be rolled out, as data would be lost.
+The first case - double scraping - should at minimum cause a warning.
+
+## Required configuration options
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus custom resource definition](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus).
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'
+    mode: 'Classic'    
+
+    # Optional: array of relabel configurations to insert before the sharding
+    # relabel configuration.
+    additionalRelabelConfig: [] 
+
+    # The following section is only valid if "mode" is set to "Classic"
+    classic:
+        # Metric label used for sharding.
+        sourceLabel: '__address__'
+      
+    # The following section is only valid if "mode" is set to "Topology"
+    # 'topology.kubernetes.io/zone' for 'topology'
+    topology: 
+        # A kubernetes node label defining the topology to be sharded on
+        nodeLabel: 'topology.kubernetes.io/zone'
+        
+        # A prometheus metric containing the topology value of a given target
+        sourceLabel: '__meta_kubernetes_pod_label_topology_kubernetes_io_zone'
```

Comment:
This allows for more flexibility and is also more future proof.
Maybe you want to scrape based on region, or some other concept. Maybe the source label is generated.
We cannot anticipate what this will look like, but we can allow overrides - and this is what is done here.

Reply:
Essentially this is coming from the comment below, that pods do not have a zone label by default. If this label is injected by the user, we don't know the name.

However, it depends a bit on if and how we can solve the issue of the missing label on pods in general.

Reply:
> However, it depends a bit on if and how we can solve the issue of the missing label on pods in general.

Now I remember that I had an idea which avoids the need to attach node metadata to the service discovery. As explained in https://gist.github.com/simonpasquier/0a7df5b0899721167b5c56f9cdc9668c (which was mentioned in https://github.com/prometheus-operator/prometheus-operator/issues/5495), the operator could discover the list of nodes + AZs and inject a relabel config rule which would filter the pods based on their `__meta_kubernetes_pod_node_name` label.

Reply:
@simonpasquier @nicolastakashi I added a new section to discuss this topic to the document.

Reply:
> I'm afraid this method will cause issues, as the operator would have to constantly update the relabel config. Especially in clusters where nodes scale up and down a lot, this would put quite a bit of load on prometheus and the operator.

I'm not too worried about the additional load unless the node churn is very high. Attaching node metadata to all targets has also a cost for Prometheus and the Kubernetes API (more resources being watched). I still think that it would make sense as an option for clusters which don't expect a high churn on nodes.

Reply:
But then you still have the race condition, don't you?

BTW - I just saw that prometheus 3 has `__meta_kubernetes_endpointslice_endpoint_zone` which means we have that information for serviceMonitors, right?

Reply:
Continued in a [discussion below](https://github.com/prometheus-operator/prometheus-operator/pull/7110#discussion_r1856872505)